File: backend/app/schemas/referral.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ReferralCreate(BaseModel):
    source: Literal["profile", "settings", "community", "other"] | None = None


class ReferralRedeem(BaseModel):
    code: str = Field(min_length=4, max_length=64)


class ReferralInviteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    inviter_id: int
    accepted_user_id: int | None = None
    status: Literal["PENDING", "REGISTERED", "EXPIRED", "REVOKED"]
    reward_status: Literal["NOT_APPLICABLE", "PENDING", "GRANTED"] | None = None
    created_at: datetime
    registered_at: datetime | None = None
    expires_at: datetime | None = None
    invite_url: str | None = None


class ReferralMeResponse(BaseModel):
    code: str
    invite_url: str
    invite: ReferralInviteResponse | None = None


class ReferralStatsResponse(BaseModel):
    invitations_sent: int
    successful_registrations: int


class ReferralRedeemResponse(BaseModel):
    redeemed: bool
    invite: ReferralInviteResponse

File: packages/shared/src/schemas/referral.ts
import { z } from "zod";

export const ReferralInviteStatusSchema = z.enum([
  "PENDING",
  "REGISTERED",
  "EXPIRED",
  "REVOKED",
]);

export const ReferralRewardStatusSchema = z.enum([
  "NOT_APPLICABLE",
  "PENDING",
  "GRANTED",
]);

export const ReferralInviteSchema = z.object({
  id: z.number().int().positive(),
  code: z.string(),
  inviter_id: z.number().int().positive(),
  accepted_user_id: z.number().int().positive().nullable().optional(),
  status: ReferralInviteStatusSchema,
  reward_status: ReferralRewardStatusSchema.optional(),
  created_at: z.string().datetime(),
  registered_at: z.string().datetime().nullable().optional(),
  expires_at: z.string().datetime().nullable().optional(),
  invite_url: z.string().url().optional(),
});

export const ReferralMeSchema = z.object({
  code: z.string(),
  invite_url: z.string().url(),
  invite: ReferralInviteSchema.optional(),
});

export const ReferralStatsSchema = z.object({
  invitations_sent: z.number().int().nonnegative(),
  successful_registrations: z.number().int().nonnegative(),
});

export const ReferralCreateSchema = z.object({
  source: z.enum(["profile", "settings", "community", "other"]).optional(),
});

export const ReferralRedeemSchema = z.object({
  code: z.string().trim().min(4).max(64),
});

export const ReferralRedeemResponseSchema = z.object({
  redeemed: z.boolean(),
  invite: ReferralInviteSchema,
});

export type ReferralInviteStatus = z.infer<typeof ReferralInviteStatusSchema>;
export type ReferralRewardStatus = z.infer<typeof ReferralRewardStatusSchema>;
export type ReferralInvite = z.infer<typeof ReferralInviteSchema>;
export type ReferralMe = z.infer<typeof ReferralMeSchema>;
export type ReferralStats = z.infer<typeof ReferralStatsSchema>;
export type ReferralCreate = z.infer<typeof ReferralCreateSchema>;
export type ReferralRedeem = z.infer<typeof ReferralRedeemSchema>;
export type ReferralRedeemResponse = z.infer<typeof ReferralRedeemResponseSchema>;
